src/bin/driver.rs

```diff
@@ -419,6 +435,81 @@ async fn run_level_last(
     Ok(())
 }
 
+async fn run_aggregate_by_attributes(
+    cfg: &config::Config,
+    mastic: &MasticHistogram,
+    client_0: &CollectorClient,
+    client_1: &CollectorClient,
+    attributes: &[String],
+    num_clients: usize,
+) -> io::Result<()> {
+    for start in (0..num_clients).step_by(cfg.flp_batch_size) {
+        let end = std::cmp::min(num_clients, start + cfg.flp_batch_size);
+        let req = AggregateByAttributesValidateRequest {
+            attributes: attributes.to_vec(),
+            start,
+            end,
+        };
+
+        // For each report, each aggregator evaluates the VIDPF on each of the attributes and returns
+        // the VIDPF proof and its FLP verifier share.
+        let t = Instant::now();
+        let resp_0 = client_0.aggregate_by_attributes_start(long_context(), req.clone());
+        let resp_1 = client_1.aggregate_by_attributes_start(long_context(), req.clone());
+        let (results_0, results_1) = try_join!(resp_0, resp_1).unwrap();
+        assert_eq!(results_0.len(), req.end - req.start);
+        assert_eq!(results_1.len(), req.end - req.start);
+
+        let mut rejected = Vec::with_capacity(req.end - req.start);
+        for (client_index, ((mut verifier, eval_proof_0), (verifier_share_1, eval_proof_1))) in
+            (req.start..req.end).zip(results_0.into_iter().zip(results_1.into_iter()))
+        {
+            vec_add(&mut verifier, &verifier_share_1);
+            if !mastic.decide(&verifier).unwrap() {
+                println!("Report {client_index} rejected (weight check failed)");
+                rejected.push(client_index);
+            }
+
+            if eval_proof_0 != eval_proof_1 {
+                println!("Report {client_index} rejected (onehot or path check failed)");
+                rejected.push(client_index);
+            }
+        }
+
+        println!(
+            "{start}..{end}: report validation completed in {:?}: rejected {} reports",
+            t.elapsed(),
+            rejected.len()
+        );
+
+        let t = Instant::now();
+        let req = AggregateByAttributesResultRequest {
+            rejected,
+            num_attributes: attributes.len(),
+            start,
+            end,
+        };
+
+        let resp_0 = client_0.aggregate_by_attributes_finish(long_context(), req.clone());
+        let resp_1 = client_1.aggregate_by_attributes_finish(long_context(), req.clone());
+        let (mut results, results_share_1) = try_join!(resp_0, resp_1).unwrap();
+        for (r, s1) in results.iter_mut().zip(results_share_1.iter()) {
```

Comment:
you can probably do par_iter here

Reply:
well, we're doing a reduction here, so parallelizing would be complicated. This is just a few hundred microseconds anyway.